ext/include/opentelemetry/ext/http/client/curl/http_client_curl.h

```diff
@@ -143,19 +143,19 @@ class Session : public opentelemetry::ext::http::client::Session
   }
 
   virtual void SendRequest(
-      opentelemetry::ext::http::client::EventHandler &callback) noexcept override
+      std::shared_ptr<opentelemetry::ext::http::client::EventHandler> callback) noexcept override
```

Comment:
Can we transfer the callback ownership, if that helps us avoid shared_ptr?

Reply:
Moving the callback ownership to http_client, would involve storing the unique_ptr to some kind of map (similar to session) and clearing the entry in the map when the async thread completes. But I could not figure out the simplest way to let know http_client that the async thread is completed and map needs to be cleaned. However, cleanup could be achieved by running a separate cleanup thread which cleans up on getting notified by async thread.

Do you have anything else in mind?

The other approach I was thinking was to move the ownership to the lambda function defined in Session::SendRequest, because this lambda function would run in async thread and callback should be valid till lambda function ends. This could have been doable in C++14 as it supports std::move in lambda capture. But I could not think of achieving the same in C++11. 

Also, since we have maintained two versions of Export implementation in es_log_exporter.cc, the scope of Export outlives the async thread and in another version the async thread outlives the scope of Export, unique_ptr needs to outlive both of them.

Let me know your thoughts.

Reply:
I was thinking whether to add a background thread to deal with resources of async HTTP call?
We will also need a thread to run [`curl_multi_perform`](https://curl.se/libcurl/c/[curl_multi_perform](https://curl.se/libcurl/c/curl_multi_perform.html).html) when I start to implement the real async version of `curl::Session::SendRequest`.
With a background worker thread. I think we can allocate the unique_ptr<EventHandler> on stack in sync Export and move it into the background worker thread in the async Export.

> This could have been doable in C++14 as it supports std::move in lambda capture. But I could not think of achieving the same in C++11.
 We can use parameter to move unique_ptr in C++11. Here is a example https://godbolt.org/z/T5sG7K9fn

Reply:
> We can use parameter to move unique_ptr in C++11. Here is a example https://godbolt.org/z/T5sG7K9fn

Thanks @owent  for sharing this.

Reply:
> Also, since we have maintained two versions of Export implementation in es_log_exporter.cc, the scope of Export outlives the async thread and in another version the async thread outlives the scope of Export, unique_ptr needs to outlive both of them.

@owent But according to above reasoning, I think we can start with shared_ptr for now.